docs/caat.1
.\" Copyright 2015 Chris Taylor
.\"
.\" This file is part of caatinga.
.\"
.\" Caatinga is free software: you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation, either version 3 of the License, or
.\" (at your option) any later version.
.\"
.\" Caatinga is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with caatinga.  If not, see <http://www.gnu.org/licenses/>.
.\"
.\" Man page for the caat program
.\"
.TH caat 1 "June 14 2014" 2.0 caat


.SH NAME
.B caat
\- create a full system backup.


.SH SYNOPSIS
.B caat
[options]


.SH DESCRIPTION
.B caat
is a backup program used for making full system backups.  Each time a backup is
performed, a new snapshot of the file system is created.  This creates a history
of all files on your system allowing you to restore files from any point in
time.
.B caat
uses hard links when creating snapshots.  As a result, each backup takes a
minimal amount of disk space and time to execute.  This efficiency allows
backups to be ran frequently in a command scheduler to insure current data is
always backed up.

Each backup that's made is organized by the systems host name.  This allows more
than one system to use the same backup location.  Files that are backed up can be
viewed, restored or deleted using the
.BR lscaat (1)
tool.
.B caat
also uses an external configuration file called
.BR caatinga.conf (5)
which includes all information needed to perform a system backup.  This eliminates
the need to remember numerous cryptic command options to backup your system, simply
run
.B caat
and it does it's job.


.SH OPTIONS
.TP
.BR \-b " location, " \-\-backup\-location =<location>
Backup location to use.  This will override the setting in caatinga.conf.
.TP
.BR \-\-clean
Manually remove backups marked for deletion.
.TP
.BR \-c " file, " \-\-config =<file>
Specify an alternate configuration file.
.TP
.BR \-d ", " \-\-delete\-oldest
Delete the oldest backup for this system.
.TP
.BR \-g ", " \-\-register
Register the backup location for use by this machine.
.TP
.BR \-h ", " \-\-help
Displays help message.
.TP
.BR \-n " hostname, " \-\-hostname =<hostname>
Use this hostname instead of what is defined for the local system.
.TP
.BR \-r " path, " \-\-root =<path>
Specify an alternate root filesystem path.
.TP
.BR \-v ", " \-\-verbose
Verbose mode.  Display backup activity.
.TP
.BR \-V ", " \-\-version
Displays version information and exits.


.SH FILES
.I /etc/caatinga/caatinga.conf
.RS
The configuration file used by
.B caat
and
.B lscaat.
See
.BR caatinga.conf (5)
for further details.


.SH EXAMPLES
.P
Register the backup location in
.BR caatinga.conf (5)
as a backup device.
.RS
.P
.B caat
\-\-register
.RE

.P
Create a backup of the current system.
.RS
.P
.B caat
.RE

.P
Create a backup of the current system using an alternate backup location.
.RS
.P
.B caat
\-\-backup\-location=/mnt/sdg1
.RE

.P
Create a backup of another file system specifying an alternate host name.
.RS
.P
.B caat
\-\-root=/mnt/sdb1 \-\-hostname=casper
.RE

.SH AUTHOR
Chris Taylor <headmastersquall at gmail dot com>


.SH SEE ALSO
.BR lscaat (1)
.BR caatinga.conf (5)


docs/ChangeLog
??? - Next Release

  - Changed lock file name to not use the execuatble name.


1.1.1 - 05/21/2015

  - Fixed wrong default hook directories in the sample file.


1.1.0 - 05/21/2015

  - Moved lock file to /tmp and its name includes the host name.
  - Do not allow caat -d to delete the oldest if ony one backup exists.
  - Do not show ignored files on lscaat changes output.
  - Renamed pre_hooks to pre_backup_hooks from caatinga.conf.
  - Renamed post_hooks to post_backup_hooks from caatinga.conf.
  - Added pre_restore_hooks to caatinga.conf.
  - Added post_restore_hooks to caatinga.conf.
  - Fixed locking bug that would corrupt a running backup if another instance
    attempted to run.
  - Added check to see if the locked pid is actually running.
  - Added max_days to caatinga.conf.
  - Added max_images to caatinga.conf.
  - Converted CommandArgs to use argparse and renamed to args.
  - Normalized all module names to be lower case.
  - Removed drive_percentage from caatinga.conf.  Replaces with max_images
    and max_days.  
  - Removed requirement of having the backup destination a mounted device.
  - Reworked setup.py to use setuptools instead of disttools.


caat.py
#!/usr/bin/env python

# Copyright 2012 Chris Taylor
#
# This file is part of caatinga.
#
# Caatinga is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# Caatinga is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with caatinga.  If not, see <http://www.gnu.org/licenses/>.

import os
import caatinga.core.errorcodes as errorcodes
import caatinga.core.functions as fn
import caatinga.caat.backup as backup
from sys import argv
from time import strftime
from caatinga.caat.organizer import organize
from caatinga.core.CommandArgs import CommandArgs
from caatinga.core.SettingsValidator import SettingsValidator

__version__ = "caatinga version: 1.0.0"
usage = """Usage: caat [options]

These options are also available in lscaat.

Options:

    -b location, --backup-location=<location>
        Backup location to use.  This will override the setting in
        caatinga.conf.

    -c file, --config=<file>
        Specify an alternate configuration file.

    -d, --delete-oldest
        Delete the oldest backup that caat has created.

    -g, --register-backup
        Register the backup location as a backup device.

    -h, --help
        Displays this help message.

    -n hostname, --hostname=<hostname>
        Use this hostname instead of what is defined for the local system.

    -r path, --root=<path>
        Use this as the root directory to be backed up.

    -v, --verbose
        Verbose mode.  Display backup activity.

    -V, --version
        Displays version information and exits.
"""


def main(args):
    """
    Application entry point.
    """
    try:
        run_backup(args)
        exit(0)
    #except OSError as er:
    #    if er.errno == errorcodes.FILESYSTEM_FULL:
    #        print("Backup drive is full")
    #    elif er.errno == errorcodes.PERMISSION_DENIED:
    #        print("Permission Denied")
    #    else:
    #        print("Operation not permitted on the mounted backup file system")
    #    exit(er.errno)
    except KeyboardInterrupt:
        exit(1)
    #except Exception as ex:
    #    print(str(ex).strip("'"))
    #    exit(1)


def run_backup(args):
    """
    Main method that performs the backup.
    """
    commandArgs = CommandArgs(args)
    if commandArgs.help:
        print(usage)
        exit(0)
    if commandArgs.version:
        print(__version__)
        exit(0)

    settings = fn.getSettingsInstance(commandArgs)
    checkForRegisterOption(settings, commandArgs)
    SettingsValidator().validate(settings)
    writer = fn.getOutputWriter(commandArgs.verbose)
    backupHome = fn.getBackupHome(settings.backupLocation, settings.hostName)
    previousBackup = os.path.realpath(fn.getLatestLink(backupHome))
    lockFile = getLockFile(backupHome)
    checkForDeleteOldest(commandArgs, backupHome)
    backupRoot = backup.createBackupRoot(
        backupHome,
        strftime("%Y-%m-%d-%H%M%S") + ".part",
        settings.backupgid)
    markPartialBackupForDeletion(backupHome)

    try:
        lock(lockFile)
        backup.backupDirectory(
            backupRoot,
            previousBackup,
            settings.root,
            settings,
            writer)
        os.rename(backupRoot, backupRoot.replace(".part", ""))
        fn.updateLatestLink(backupHome)
    finally:
        backup.removeLockFile(lockFile)

    checkForReduceBackups(settings.reduceBackups, backupHome)
    deleteBackupsMarkedForDeletion(backupHome, writer)
    checkDrivePercentage(backupHome, settings.drivePercentage, writer)


def checkForRegisterOption(settings, commandArgs):
    """
    Check the command args for the register option and register the backup
    device if provided.  This will exit with a status 0 if a device was
    registered.
    """
    if commandArgs.register:
        fn.registerBackupDevice(
            settings.backupLocation,
            settings.backupgid)
        print(("The device mounted at {0} is now registered as a " +
              "backup device.").format(settings.backupLocation))
        exit(0)


def getLockFile(backupHome):
    """
    Returns the name of the lock file based on the executable name.
    """
    exe = os.path.basename(argv[0]).rstrip(".py")
    return os.path.join(backupHome, exe + ".pid")


def checkForDeleteOldest(commandArgs, backupHome):
    """
    Check if the option to delete the oldest backup was provided, and delete
    the oldest backup.
    """
    if commandArgs.deleteOldest:
        fn.deleteBackup(backupHome, fn.getOldestBackup(backupHome))
        exit(0)


def markPartialBackupForDeletion(backupHome):
    """
    Looks for any partial backups and marks them for deletion.
    """
    partials = fn.getPartialBackups(backupHome)
    for partial in partials:
        partialBackup = os.path.join(backupHome, partial)
        os.rename(partialBackup, partialBackup.replace(".part", ".delete"))


def lock(lockFile):
    """
    Creates a lock file if one is not present.  If a lock file already exists,
    an exception will be thrown.
    """
    if os.path.exists(lockFile):
        with open(lockFile) as lock:
            pid = lock.readline()
        raise Exception("A backup is currently running [{0}]".format(pid))
    backup.createLockFile(lockFile)


def checkForReduceBackups(reduceBackups, backupHome):
    """
    Reduce the backups if the reduce backups option is provided.
    """
    if reduceBackups:
        organize(backupHome)


def deleteBackupsMarkedForDeletion(backupHome, writer):
    """
    Delete backups that are marked to be deleted.
    """
    for backup in fn.getBackupsMarkedForDeletion(backupHome):
        writer("Deleting: {0}".format(backup))
        fn.deleteBackup(backupHome, backup)


def checkDrivePercentage(backupHome, drivePercentage, outputWriter):
    '''
    Compares drive usage with user settings and deletes old backups as
    necessary.
    '''
    while drivePercentage < fn.getDriveUsagePercentage(backupHome):
        if len(fn.getBackups(backupHome)) <= 1:
            break
        oldest = fn.getOldestBackup(backupHome)
        outputWriter("Deleting: {0}".format(oldest))
        fn.deleteBackup(backupHome, oldest)

if __name__ == "__main__":
    main(argv)
